### PR TITLE
Do not throw exceptions within destructor

### PR DESCRIPTION
**Pull request by oflebbe**
Introducing a "C" krb5_error_exit() and refatcoring error_exit() infrastructure from msktutil

Should be tested. But compiles clean with gcc-7 on macosx (aside from deprecations at macosx level)

### CODE REVIEW

File: krb5wrap.h [hunk 1]
```diff
@@ -72,15 +74,16 @@ class KRB5Keytab {
 public:
     KRB5Keytab(const std::string &keytab_name) : m_keytab() {
         krb5_error_code ret = krb5_kt_resolve(g_context, keytab_name.c_str(), &m_keytab);
-        if (ret)
-            throw KRB5Exception("krb5_kt_resolve", ret);
+        if (ret) {
+            krb5_error_exit("krb5_kt_resolve",ret);
+        }
```
Comment (sckobras) [inline krb5wrap.h]: Why did you choose to replace the exception with a forced exit in the ctor as well? While not used today, being able to catch and handle a failure at this stage looks reasonable to me.

File: krb5wrap.cpp [hunk 2]
```diff
@@ -270,8 +273,7 @@ KRB5Keytab::cursor::~cursor()
                                               m_keytab.m_keytab,
                                               &m_cursor);
     if (ret) {
-        /* FIXME: shouldn't throw from destructor... */
-        throw KRB5Exception("krb5_kt_end_seq_get", ret);
+        krb5_error_exit("krb5_kt_end_seq_get", ret);
```
Comment (sckobras) [inline krb5wrap.cpp]: Exception from `KRB5Keytab::cursor` are currently caught and ignored, so a forced exit changes (intended) semantics.

**Comment by oflebbe** [comment]
@sckobras I reworked the patch to adress some of your findings. 
My intention was to remove the exception handling from the code, this as a first step. But nevertheless you are right, the ctor patch is problematic and may be plain wrong. I am continue to work in this.

**Comment by oflebbe** [comment]
Ok, this is all plain wrong. We are catching exceptions we should not (like permission denied on keytabs ). Will have to dig deeper. And we have to use exceptions, otherwise many things will break within msktutil,

**Comment by oflebbe** [comment]
Ok, this seems way better. I removed try/catches of keytab read errors. Code looks a lot more natural.
I removed scoping brackets I intruduced earlier and added a reset() function to reuse keytab cursors.
And fixed a more or less bug: m_entry was not properly initialized.

If noone complains I like to push this tomorrow

**Comment by oflebbe** [comment]
Thanks for the review @sckobras !

**Pull request merged**